src/cargo/core/compiler/custom_build.rs

```diff
@@ -562,10 +570,28 @@ impl BuildOutput {
                 "rustc-link-lib" => library_links.push(value.to_string()),
                 "rustc-link-search" => library_paths.push(PathBuf::from(value)),
                 "rustc-link-arg-cdylib" | "rustc-cdylib-link-arg" => {
+                    if !targets.iter().any(|target| target.is_cdylib()) {
+                        bail!(
+                            "invalid instruction `cargo:{}` from {}\n\
+                             The package {} does not have a cdylib target.",
+                            key,
+                            whence,
+                            pkg_descr
+                        );
+                    }
                     linker_args.push((LinkType::Cdylib, value))
                 }
                 "rustc-link-arg-bins" => {
                     if extra_link_arg {
+                        if !targets.iter().any(|target| target.is_bin()) {
+                            bail!(
+                                "invalid instruction `cargo:{}` from {}\n\
+                                 The package {} does not have a bin target.",
+                                key,
+                                whence,
+                                pkg_descr
+                            );
+                        }
```

Comment:
Can we extract the same logic from these three `bail!`?

Reply:
I thought about that, but I didn't see a way to consolidate it such that it wouldn't end up with more total lines of code. Only the first two have the same structure. The third has a different structure and text, which would need customization.  

Thanks for the suggestion, I do agree it is a bit repetitive and verbose.  Hopefully [RFC 2795](https://rust-lang.github.io/rfcs/2795-format-args-implicit-identifiers.html) will make progress, which I think would simplify this a bit.  

If there are more link-types added in the future, perhaps might spend some more effort to consolidate it.

Reply:
Got it. Thank for your detailed explanation!
BTW, RFC 2795 looks really promising 😲